### PR TITLE
first draft of handling marker exceptions wrapped in exceptiongroup

### LINKED ISSUE #4106: Our internals should handle `StopTest` and other marker exceptions even when wrapped in an `ExceptionGroup`
Hypothesis currently misbehaves if the marker exceptions we raise are wrapped up in an `ExceptionGroup`, e.g.:

```python
import trio
from hypothesis import given, strategies as st

@given(st.data())
def test(data):
    async def task(pred):
        return data.draw(st.booleans().filter(pred))

    @trio.run
    async def _():
        async with trio.open_nursery() as nursery:
            nursery.start_soon(task, bool)
            nursery.start_soon(task, lambda _: False)
```
```
  + Exception Group Traceback (most recent call last):
  |   ...
  +-+---------------- 1 ----------------
    | Traceback (most recent call last):
    |   ...
    | hypothesis.errors.StopTest: 1
    +---------------- 2 ----------------
    | Traceback (most recent call last):
    |   ...
    | hypothesis.errors.Frozen: Cannot call start_example on frozen ConjectureData
    +------------------------------------
```

In such cases I think it's pretty clear that we should treat an exception group _containing_ a `StopTest` as we would a `StopTest` itself.  I think it's also pretty clear that any `Frozen` exceptions inside an `ExceptionGroup` are causally downstream of some other exception, and can therefore be discarded so long as there's any other exception in the group too.

### PR DESCRIPTION
**Pull request by jakkdl**
I have barely used Hypothesis, and never worked with the codebase, so there are several gaps in code coverage and question marks on implementation decisions. If somebody more familiar wants to take a quick pass and opine it'd be very helpful :)

I can get code coverage by manually raising the specific `ExceptionGroup`s that'll trigger the different branches, but it'd likely be better to do the actual things that causes those exceptions. So would likewise appreciate quick pointers on how to, if at all possible, cause e.g. multiple `StopTest`, multiple `Frozen`, a single `Frozen`, etc.

fixes #4106

### CODE REVIEW

**Review by jobh — COMMENTED** [review]
Very nice, thanks! :+1: 

I think most of the build errors will go away with some of the minor commented changes, the remaining will hopefully be manageable :-)

**Comment by jobh** [comment]
> I can get code coverage by manually raising the specific `ExceptionGroup`s that'll trigger the different branches, but it'd likely be better to do the actual things that causes those exceptions. So would likewise appreciate quick pointers on how to, if at all possible, cause e.g. multiple `StopTest`, multiple `Frozen`, a single `Frozen`, etc.

It should be possible to provoke most of these situations by playing around with multiple tasks failing in different ways. But I think it's fine to create the non-obvious situations explicitly, too.

**Comment by jobh** [comment]
> It should be possible to provoke most of these situations by playing around with multiple tasks failing in different ways. But I think it's fine to create the non-obvious situations explicitly, too.

Sorry: "most" should probably be replaced by "some" :grin:

**Comment by jobh** [comment]
Oh, and one other tip: hypothesis will transform/trim the tracebacks by default, which may be confusing sometimes when working on exceptions and not seeing what you'd expect. See `get_trimmed_traceback()`.

File: hypothesis-python/tests/cover/test_exceptiongroup.py
Comment (jakkdl) [reply hypothesis-python/tests/cover/test_exceptiongroup.py:53]: ye that's probably the way to go. I'm incredibly biased on the matter as the person who wrote `trio.testing.RaisesGroup` and is trying to get it added to pytest :innocent:

Comment (Zac-HD) [reply hypothesis-python/tests/cover/test_exceptiongroup.py:53]: (and as a dual maintainer of Pytest and Trio, among other things, I'm pretty keen to support that when I have time 😅)

File: hypothesis-python/src/hypothesis/core.py
Comment (jakkdl) [reply hypothesis-python/src/hypothesis/core.py:812]: > If all exceptions are Frozen, raise the first (as above).

Actually, I think this could cause confusion. If I spawn three tasks and I only get one `Frozen`, I will assume that the-thing-that-caused-Frozen only happened in one of the tasks. We should probably err on the side of not unwrapping unless necessary to get internal exception handling to work.

Comment (jakkdl) [reply hypothesis-python/src/hypothesis/core.py:812]: Another question is if we get multiple `HypothesisException`/`StopException`, if that's possible?

**Comment by jakkdl** [comment]
> > I can get code coverage by manually raising the specific `ExceptionGroup`s that'll trigger the different branches, but it'd likely be better to do the actual things that causes those exceptions. So would likewise appreciate quick pointers on how to, if at all possible, cause e.g. multiple `StopTest`, multiple `Frozen`, a single `Frozen`, etc.
> 
> It should be possible to provoke most of these situations by playing around with multiple tasks failing in different ways. But I think it's fine to create the non-obvious situations explicitly, too.

actually, make that *none* of the situations if we 1. Want coverage to pass on py<3.11 and 2. not add a dependency on trio or anyio.
`asyncio.TaskGroup` wasn't added until py3.11, and afaik nothing in the stdlib will raise `ExceptionGroup` before that.

File: hypothesis-python/src/hypothesis/core.py
Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/core.py:814]: If we get multiple `StopTest`, re-raise the one with the lowest _testcounter i.e. the first one to occur.

**Comment by Zac-HD** [comment]
I think for testing we'll want `/cover/` tests which do whatever explicit super-weird code we need ([we have `ExceptionGroup`](https://github.com/HypothesisWorks/hypothesis/blob/4d079f7a62a6fdabfb2f2a2b276cb0f721ea62de/hypothesis-python/src/hypothesis/internal/compat.py#L22-L29)), _and also_ some `/nocover/` tests which only run on py311+ and use `asyncio.TaskGroup` in a somewhat-more-natural way.  Then when 310 goes end of life, we can promote them over to replace the weird cover tests.

**Review by Zac-HD — COMMENTED** [review]
(partial review is better than no review, right?)

I've also merged a fix for https://github.com/HypothesisWorks/hypothesis/issues/4115 so you might want to rebase on master 🙂

File: hypothesis-python/src/hypothesis/core.py
Comment (jakkdl) [reply hypothesis-python/src/hypothesis/core.py:1141]: that would only override the `__cause__` no? And the comment specifically talks about errors happening in `finally`, where you rarely raise explicit exceptions.
It seems pretty dangerous to ignore all exceptions in case the data is frozen, shouldn't the user be informed if there's errors happening in their teardown code?
Although thinking about it more, the problem is perhaps that the `StopTest.__cause__` is not getting reraised/logged?

**Comment by jakkdl** [comment]
#4115 made me realize that unwrapping exceptions inside groups is *really tricky* to do right without losing the context/cause of the original exception. Though I'm unsure if/how much this matters for hypothesis, as we only try to unwrap specific marker exceptions which users wouldn't see anyway.

File: hypothesis-python/src/hypothesis/core.py
Comment (jakkdl) [reply hypothesis-python/src/hypothesis/core.py]: do you also want to avoid it in test files?

File: hypothesis-python/src/hypothesis/core.py
Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/core.py:1151]: Yeah, I think you're correct here, we do want to expose errors in teardown code.

> Although thinking about it more, the problem is perhaps that the `StopTest.__cause__` is not getting reraised/logged?

hmm... we never `raise StopTest(...) from ...`, but we certainly do raise it from inside exception handlers that we don't want to expose the user to (via something like `except ValueError: data.mark_invalid()`).  I guess we could inspect `__cause__`, but it'd have so many missed alarms that really doesn't seem worth doing :/

Comment (jakkdl) [reply hypothesis-python/src/hypothesis/core.py:1151]: > hmm... we never raise StopTest(...) from ...

you do it on the very next line in this file :sweat_smile: 
```python
                 raise StopTest(data.testcounter) from e
```
Though I suppose saving it as `__cause__` for reraising in the engine is perhaps not the way to do it, and the caught exception should instead be saved Somewhere Else:tm:

Comment (jakkdl) [reply hypothesis-python/src/hypothesis/core.py:1151]: Oh and to be explicit, I think traversing `__context__` for `StopTest` would be robust. As far as I know there's no standard way of losing entries in it - but not terribly elegant to traverse when you also have to deal with exception groups.

I'll try to write a test and open a separate issue/PR for it

File: hypothesis-python/src/hypothesis/core.py
Comment (jakkdl) [reply hypothesis-python/src/hypothesis/core.py]: hrm, I could swear there was a flake8 plugin with an optional rule that banned `from __future__ import annotations` but can't find it (nor any ruff rule).

File: hypothesis-python/src/hypothesis/core.py
Comment (jakkdl) [reply hypothesis-python/src/hypothesis/core.py:1151]: opened #4132 to continue the discussion

**Comment by jakkdl** [comment]
idk why the 3.8 builds are failing, but #4131 will drop 3.8 support so I can hopefully just ignore that :innocent:
Otherwise this is mostly just blocked on reviews/taking decisions. There are a couple test cases, esp multiple `HypothesisException`, that would benefit from more proper exceptiongroup support, but I'm not sure if we want to handle that in this PR. E.g. an `HypothesisExceptionGroup` raised in `_execute_once_for_engine` and later handled in the engine.

**Review by Zac-HD — COMMENTED** [review]
Sorry this has taken me so long to get to 😞 

I haven't tried to answer the questions in comments in this round - better to get you some response than more delay! - but overall the design looks good to me, and if we rebase on master to drop py3.8 and it's all still green I think we're pretty close to done.

File: hypothesis-python/tests/nocover/test_exceptiongroup.py
Comment (jakkdl) [reply hypothesis-python/tests/nocover/test_exceptiongroup.py]: fixed!

**Comment by jakkdl** [comment]
I have no clue what that py313 failure is about, it doesn't look related to this PR?

File: hypothesis-python/src/hypothesis/core.py
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/core.py:739]: We might want a name that indicates _what_ we're unwrapping, since this might be confused for a generic unwrap-single-exception helper (which is usually a bad idea, but...).  Maybe `unwrap_markers_from_group`?

File: hypothesis-python/src/hypothesis/core.py
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/core.py:793]: ... I think at some point we have to declare this is an improvement on the status quo, and ship it even if it doesn't resolve every single thing we can think of - we can defer these questions to a follow-up issue, for example.

File: hypothesis-python/tests/cover/test_exceptiongroup.py
Comment (Zac-HD) [inline hypothesis-python/tests/cover/test_exceptiongroup.py:24]: I think calling `.freeze()` like this is technically an invalid use of an internal API; that seems OK for the test but it'd be good to say so in a comment to avoid future confusion.

**Review by Zac-HD — COMMENTED** [review]
Overall looks good; I'll ask some users whether this solves their problem and will be inclined to merge pretty soon if so.

We'll need a `RELEASE.rst` (minor version) to do that, everything else is green 🎊

**Comment by Zac-HD** [comment]
also note-to-self, looks like a subsequent mypy upgrade has [this test passing](https://github.com/HypothesisWorks/hypothesis/blob/2fd99dd01b9653115fa9b10bfbe906388ce35e23/whole_repo_tests/test_mypy.py#L330-L331) so we can remove the xfail 🙂

File: hypothesis-python/src/hypothesis/core.py
Comment (jakkdl) [reply hypothesis-python/src/hypothesis/core.py:793]: yeah I suspect the "proper" way to fix this would be to not do any unwrapping and instead change a bunch of exception handlers in various places to handle exception groups. But that'd probably be a pretty involved fix, and require taking decisions in all those spots on how to handle multiple marker exceptions.

File: hypothesis-python/tests/cover/test_exceptiongroup.py
Comment (jakkdl) [reply hypothesis-python/tests/cover/test_exceptiongroup.py:24]: oh interesting. I thought if I stuck to non-underscored methods everywhere I'd be using public APIs, maybe something should be (deprecated and) renamed to make it clearer for users?

**Comment by jakkdl** [comment]
`whole_repo_tests/test_mypy.py::test_stateful_target_params_return_type[int-targets=(b1, b2)-initialize]` is failing, no clue why. None of the edits in my last commit should've affected runtime behavior or typing at all.

**Comment by Zac-HD** [comment]
> `whole_repo_tests/test_mypy.py::test_stateful_target_params_return_type[int-targets=(b1, b2)-initialize]` is failing, no clue why.

Unrelated flake, unfortunately - it's xfailed on master, unclear why that's still failing here...

**Comment by tybug** [comment]
`targets=(b1,)` is xfailed on master but not `targets=(b1, b2)`. I saw this flake in https://github.com/HypothesisWorks/hypothesis/pull/4169 as well, should we xfail it on the same principle?

**Review by Zac-HD — APPROVED** [review]
Alright, I think this is ready to merge!  

Thanks again to @jakkdl for the implementation, and @jobh and @tybug for helping with review 😍

File: hypothesis-python/tests/nocover/test_exceptiongroup.py
Comment (jakkdl) [inline hypothesis-python/tests/nocover/test_exceptiongroup.py:215]: this perhaps stands out as something to clean up/move elsewhere before merge, but otherwise I think it's mergeable

Comment (Zac-HD) [reply hypothesis-python/tests/nocover/test_exceptiongroup.py:215]: I've just dug through, and I'm pretty sure it's impossible to hit this with public API - we only freeze the conjecturedata object after the user's code is done.  On that basis I'd be happy to just delete these tests and merge 🙂

**Pull request merged**
